### PR TITLE
Cleaned up MemleakTest and expanded its use in the tests

### PR DESCRIPTION
**Pull request by mnmelo**
Most tests in `test_atomgroup.py`, `test_coordinates.py`, and `test_persistence.py` are now checked for memleaks. This should cover most of the candidate cases for memleaks (Universes and Readers). As far as I could tell there was no slowdown in testing.

Renamed private members of `MemleakTest` to underscored names. Improved the class docs.

Moved the persistent offset tests to the newly-created `test_persistence.py`.

### CODE REVIEW

**Comment by richardjgowers** [comment]
So testing all objects for memory leaking is a good idea.  I don't like putting `MemleakTest` into all the other tests though (especially in `test_atomgroup`), it's fairly complicated code, although probably benign.  Plus we don't need to check for memory leaking every time we create an object, we just need targeted tests?

Could we instead create a set of light tests that only have one purpose

So...

``` python
class TestAtomGroupPersistence(MemleakTest):
    def test_atomgroup(self):
        ag = u.atoms[:10]
        assert_equal(len(ag), 10)
```

Then repeat this pattern for all other objects? (`AtomGroup`, `Atom`, `Residue` etc)

That said, I do like putting it in for each Reader/Writer test in `test_coordinates`, so maybe keep it in there.

**Comment by mnmelo** [comment]
Yup, I agree that monkey-patching everything seems a bit excessive.
What we could do is leave MemleakTest off for the moment, and I'll try to also implement something along the lines of option 1 of issue #323 (the one that only tests for memleaks once, but breaks parallel testing). I'll try to have it kick in only when a serial testing run is detected.

Then the development workflow would become:
0- Make changes to the code;
1- Test locally in parallel; 
2- Repeat steps 0 and 1 until all passes;
3- Optionally run tests in serial one final time before committing (this triggers tests for leaks once);
4- Commit/pull request and let Travis build it. This should also trigger a single leak test for the entire run (that's why step 3 is optional);
5- If a leak is detected in steps 3 or 4 then dig in in more detail and use `MemleakTest` on each of the tests that cover the code changes in step 0.

This way we get to have our cake and eat it too. What say you?

**Comment by orbeckst** [comment]
I agree with the previous discussion that we only need to test the major classes once for memleaks. This can still be done in parallel, can't it? Why not add specific (parallel) memleak tests for `Universe`, `Atom`, `AtomGroup` and the Readers/Writers?

In principle I have no problems with having additional serial tests. Admittedly, I have quickly come to rely on the "hack on feature branch, PR, wait for travis" pattern (with the occasional targeted `nosetests -v test_whatever_i_am_hacking.py`).

Is my understanding correct that we gain by executing the "global memleak test" knowledge of any class that leaks, even if we did not specifically test it?

On a somewhat related note to this PR: I can see the logic of sequestering the persistence and memleak tests into their own `test_persistence.py` file (together with the offsets) – and I might even have suggested this a while back. However, I fear that this will get increasingly confusing and people will have a hard time deciding where tests go, especially when e.g. new readers are added. Instead, for a heavily object-oriented library like MDAnalyis, it might be more straightforward to partition tests by classes. Although I should really raise a new issue for this, I was even thinking about splitting the unwieldy big `test_coordinates.py` into separate per-format test files (possibly in a subdirectory `coordinates` that pretty much mimics the library layout), and `test_atomgroup.py` is another candidate.

So in summary, I'd say go ahead with

1. testing memleaks for individual classes (should still run in parallel)
2. add a global memleak test 
3. start collecting tests in `test_persistence` to get moving but be prepared to refactor.

(Oh, and fix the issues that make [travis #161 fail](https://travis-ci.org/MDAnalysis/mdanalysis/builds/68091011).)

**Comment by mnmelo** [comment]
So, I've spent quite some time on it and:
It's close to impossible to access nose config flags from within the test suite (if someone knows otherwise please let me know!). This means that I can't activate the global test (at module teardown) based on whether the test run is parallel.

This isn't a show-stopper because we can let nose's multiprocess plugin know if a given fixture can be run in parallel (is "reentrant"), and I've been able to use this to get the global memleak test work in parallel.
I've since hit a block where, like for the individual memleak test, at module teardown many objects are still in scope, and won't be reported because `gc` doesn't know yet about their uncollectable nature. The solution will likely be as done for the individual cases, where a list of pre-test objects is compared to post-test ones and the difference is del'd. I just have to fiddle more around to find out exactly which scope `teardown_module` sees.

I've also looked into writing a nose plugin class, which seems to not be that hard. Then I'd have all access to flags and whatnot. With this approach we could even have a flag, say, `--mda-memleak`, to which we could pass `global`, `individual`, or leave out for no memleak testing.
(Incidentally, a plugin may also be the way to go to get `K` to print for known failures.)

**Comment by mnmelo** [comment]
Replaced by pull request #340.
